src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointConstraintEndpointSelector.cs

```diff
@@ -72,31 +82,60 @@ public Endpoint SelectBestCandidate(HttpContext context, IReadOnlyList<Endpoint>
             }
         }
 
-        private IReadOnlyList<Endpoint> EvaluateEndpointConstraints(
+        private IReadOnlyList<EndpointSelectorCandidate> EvaluateEndpointConstraints(
             HttpContext context,
-            IReadOnlyList<Endpoint> endpoints)
+            CandidateSet candidateSet)
         {
             var candidates = new List<EndpointSelectorCandidate>();
 
             // Perf: Avoid allocations
-            for (var i = 0; i < endpoints.Count; i++)
+            for (var i = 0; i < candidateSet.Count; i++)
```

Comment:
Let's all sit together and remark how far we've come from a world where it was considered **worth it** to point out that we were using a `for` loop over `foreach` on purpose. 

This code was written by me in MVC like a zillion years ago.